### PR TITLE
Add Grid Document spec

### PR DESCRIPTION
**Pull request by shannynalayna**
Signed-off-by: Shannyn Telander <telander@bitwise.io>

### CODE REVIEW

File: community/planning/grid_document.md [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+#### Relative path in folder names
+
+Support for multi-level folder schemes could be incorporated by enforcing a
+format to indicate multiple folders in a path. Users are accustomed to folder
+and file paths joined by the backward slash character, "/", which makes it
+simple for users to understand within Grid Document.
+
+The folder state object will be extended to include a list of folder names to
+support folder depth greater than one. Grid Document would use the common
+backward slash character, "/", to indicate nested folder references. The
```
Comment (jsmitchell) [inline community/planning/grid_document.md:534]: s/backward//

File: community/planning/grid_document.md [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+Grid Document does not initially enforce permissions, but may be extended to
+support both Grid Pike and Workflow. In both cases, Grid Document could validate
+permissions to access, create, or delete a file or folder. An organization would
+be indicated as the owner of a file or folder. The organization indicated as the
+owner defines if and how agents, within and outside of the organization, may
+interact with files and folders in state.
+
+#### Incorporating Pike
+
+Grid Document can validate permissions to access, create or delete a file or
```
Comment (jsmitchell) [inline community/planning/grid_document.md:591]: access is a read operation and is not controlled by the smart contract (mutation only).

File: community/planning/grid_document.md [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+state. If the submitting agent does not have either of these permissions, the
+transaction is invalid.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This change will increase the size of state for grid instances. There is not a
+way around this as the files have to be saved somewhere. This necessitates the
+ability to groom state, i.e., deleting files and folders. However, based on
+Grid’s storage backend, this may cause issues with large amounts of historical
```
Comment (jsmitchell) [inline community/planning/grid_document.md:754]: The removal of the data from the head of the trie with delete transactions coupled with state pruning should allow complete management of this.

File: community/planning/grid_document.md [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+data being persisted.
+
+## Rationale and alternatives
+[alternatives]: #alternatives
+
+### Rationale
+
+* Grid Document allows for secure document sharing
```
Comment (jsmitchell) [inline community/planning/grid_document.md:762]: We should use a different word than "secure" here. The security domain of the transactions and state are defined by the underlying ledger. Grid is not adding or removing "security" to these record types.

File: community/planning/grid_document.md [hunk 5]
```diff
@@ -0,0 +1,808 @@
...
+* Light-weight implementation
+
+* Very familiar and simple use-case to ease users in to Grid
+
+### Alternatives
+
+* Users may use e-mail or online document sharing sites. This does not
+  necessarily provide the same amount of security as Grid Document.
+
+* Enforce permissioning, as seen in other smart contracts. This proves to be a
```
Comment (jsmitchell) [inline community/planning/grid_document.md:773]: I'm not sure this is in the right section - i.e. this is not an alternative to Grid Doc.

File: community/planning/grid_document.md [hunk 6]
```diff
@@ -0,0 +1,808 @@
...
+[prior-art]: #prior-art
+
+Interplanetary File System + good key management does largely the same thing, in
+largely the same way. It uses the hash of a file as a content ID and the file
+gets split up into a bunch of smaller cacheable pieces that get stored on other
+nodes. This caching mechanism means files can be very persistent, but by default
+they are globally accessible and unencrypted.
+
+DropBox has DocSend, which is advertised as a way to “Securely send critical
+documents and get real-time analytics”. The biggest downside is that it is close
```
Comment (jsmitchell) [inline community/planning/grid_document.md:786]: s/close/closed/

File: community/planning/grid_document.md [hunk 7]
```diff
@@ -0,0 +1,808 @@
...
+source, paid, and not easily automatable.
+
+All large software corporations have developed some sort of file sharing system,
```
Comment (jsmitchell) [inline community/planning/grid_document.md:789]: Recommend deleting this paragraph

**Review by jsmitchell — CHANGES_REQUESTED** [review]

File: community/planning/grid_document.md [hunk 8]
```diff
@@ -0,0 +1,808 @@
...
+
+### Entities
+
+Grid Document persists and allows access to local files as opaque objects,
+organized into folders.
+
+A file is referenced using its name and its parent folder name. The contents of
+the file, which are encoded by the client, are stored in state as bytes. This
+means files are stored securely and the contents are not read by the smart
+contract. Clients will handle all encoding and decoding of the file to ensure
```
[the remark this answers is not in the document]
Comment (jsmitchell) [reply community/planning/grid_document.md:41]: See below for more context on this comment.

File: community/planning/grid_document.md [hunk 9]
```diff
@@ -0,0 +1,808 @@
...
+slash character to represent the successive levels of directories. Folders in
+state would remain in a single layer. This would also use the same addressing
+formulas defined earlier and does not add complexity to storing additional
+folder contents.
+
+#### Relative path in folder names
+
+Support for multi-level folder schemes could be incorporated by enforcing a
+format to indicate multiple folders in a path. Users are accustomed to folder
+and file paths joined by the backward slash character, "/", which makes it
```
[the remark this answers is not in the document]
Comment (shannynalayna) [reply community/planning/grid_document.md:529]: can never tell between forward and backward :/

File: community/planning/grid_document.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (shannynalayna) [reply community/planning/grid_document.md:41]: Fair, it felt like a tenuous argument writing it out, but nobody commented. Will amend the description.

File: community/planning/grid_document.md [hunk 10]
```diff
@@ -0,0 +1,808 @@
...
+#### FolderCreateAction
+
+FolderCreateAction creates a new unique folder in state.
+
+```
+message FolderCreateAction {
+  string name = 1;
+}
+```
+Validation requirements:
```
[the remark this answers is not in the document]
Comment (shannynalayna) [reply community/planning/grid_document.md:171]: Would this be encapsulated by stating "Name must not contain special characters" ?

File: community/planning/grid_document.md [hunk 11]
```diff
@@ -0,0 +1,808 @@
...
+```
+$ grid doc cp \
+     local_file_name.txt \
+     remote::/documents
+```
+The above command will attempt to copy the local file, `local_file_name.txt` to
+a remote destination, `/documents`, in Grid Document state.
+
+```
+$ grid doc cp \
```
[the remark this answers is not in the document]
Comment (shannynalayna) [reply community/planning/grid_document.md:307]: @vaporos may be able to answer this question

File: community/planning/grid_document.md (hunk 2 above)
Comment (shannynalayna) [reply community/planning/grid_document.md:591]: Will remove mentions of read/access perms

File: community/planning/grid_document.md [hunk 12]
```diff
@@ -0,0 +1,808 @@
...
+“621dee” + “07” + “01” + Sha512(folder_name)[10:] + “01” + Sha512(folder_name)[48:]
+```
+
+The remaining namespaces may be defined in the future, depending on the needs of
+the smart contract or specific implementations.
+
+### Access to Grid Document state
+
+Sharing documents securely is useful alongside other Grid smart contracts. A
+smart contract can have read or write access to folders and files using Grid
```
[the remark this answers is not in the document]
Comment (shannynalayna) [reply community/planning/grid_document.md:574]: Ok, will remove mentions of the read/write access thing

File: community/planning/grid_document.md (hunk 3 above)
Comment (shannynalayna) [reply community/planning/grid_document.md:754]: Will add a sentence to explain this.

File: community/planning/grid_document.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (shannynalayna) [reply community/planning/grid_document.md:170]: Added this sentence to transactions adding a "name" field

File: community/planning/grid_document.md [hunk 13]
```diff
@@ -0,0 +1,808 @@
...
+```
+$ grid doc rm \
+     /invoices
+```
+
+Example 1:
+
+```
+$ grid doc rm \
+  a_real_folder/a_real_file
```
[the remark this answers is not in the document]
Comment (shannynalayna) [reply community/planning/grid_document.md:408]: I'm not sure, again @vaporos may be able to answer

**Review by peterschwarz — APPROVED** [review]

**Review by agunde406 — APPROVED** [review]

**Review by vaporos — APPROVED** [review]

**Pull request merged**